pkg/controllers/management/clusterprovisioner/helm_cleaner.go

```diff
@@ -0,0 +1,277 @@
+			}
+
+			if !replaced {
+				logrus.Infof("[cleanupHelmReleases] release %v in namespace %v has no deprecated or removed APIs", lastRelease.Name, namespace)
+				continue
+			}
+
+			if err := updateRelease(lastRelease, modifiedManifest, actionConfig); err != nil {
+				logrus.Errorf("[cleanupHelmReleases] failed to update release %v in namespace %v, skipping...", lastRelease.Name, lastRelease.Namespace)
+			}
+		}
+	}
+
+	return nil
+}
+
+// ReplaceManifestData replaces the out-of-date APIs with their respective valid successors, or removes an API that
+// does not have a successor.
+// Logic extracted from https://github.com/stormqueen1990/helm-mapkubeapis/blob/0245b7a7837a36fd164d83e496c453811d62c083/pkg/common/common.go#L81-L142
+func ReplaceManifestData(mapMetadata *mapping.Metadata, manifest string, kubeVersion string) (bool, string, error) {
+	var replaced = false
+	for _, mappingData := range mapMetadata.Mappings {
+		deprecatedAPI := mappingData.DeprecatedAPI
+		supportedAPI := mappingData.NewAPI
+		var apiVersion string
+
+		if mappingData.DeprecatedInVersion != "" {
+			apiVersion = mappingData.DeprecatedInVersion
+		} else {
+			apiVersion = mappingData.RemovedInVersion
+		}
+
+		if !semver.IsValid(kubeVersion) {
+			return replaced, "", errors.Errorf("Invalid format for Kubernetes semantic version: %v", kubeVersion)
+		}
+
+		if !semver.IsValid(apiVersion) {
+			return replaced, "", errors.Errorf("Failed to get the deprecated or removed Kubernetes version for API: %s", strings.ReplaceAll(deprecatedAPI, "\n", " "))
+		}
+
+		if count := strings.Count(manifest, deprecatedAPI); count > 0 {
+			if semver.Compare(apiVersion, kubeVersion) > 0 {
+				logrus.Debugf("The following API does not require mapping as the "+
+					"API is not deprecated or removed in Kubernetes '%s':\n\"%s\"\n", apiVersion,
+					deprecatedAPI)
+			} else {
+				if supportedAPI == "" {
+					logrus.Debugf("Found %d instances of deprecated or removed Kubernetes API:\n\"%s\"\n", count, deprecatedAPI)
+
+					for repl := 0; repl < count; repl++ {
+						// find the position where the API header is
+						apiIndex := strings.Index(manifest, deprecatedAPI)
+
+						// find the next separator index
+						separatorIndex := strings.Index(manifest[apiIndex:], "---\n")
+
+						// find the previous separator index
+						previousSeparatorIndex := strings.LastIndex(manifest[:apiIndex], "---\n")
+
+						/*
+						 * if no previous separator index was found, it means the resource is at the beginning and not
+						 * prefixed by ---
+						 */
+						if previousSeparatorIndex == -1 {
+							if apiIndex == 0 {
+								previousSeparatorIndex = 0
+							} else {
+								previousSeparatorIndex = apiIndex - 1
+							}
+						}
+
+						if separatorIndex == -1 { // this means we reached the end of input
+							manifest = manifest[:previousSeparatorIndex]
+						} else {
+							manifest = manifest[:previousSeparatorIndex] + manifest[separatorIndex+apiIndex:]
```

Comment:
I'm not quite sure that this part is right, though I may be missing something. 

In the following example:
```yaml
---
apiVersion: test.k8s.io/v1beta1
kind: testKind
---
```
The following values should be set:
- `previousSeparatorIndex` = 0
- `apiIndex` = 4 (3 --- and a newline)
- `separatorIndex` = 47 (len(firstLine) + len(secondLine) + 2 new lines).

In this case, the we would attempt to grab everything after index 51. I don't think that's what we want, and could produce an error. I think a similar error would also occur even if this is in the middle of the file. 

Is this something that we want to revisit, or is this something that we want to revisit?

Reply:
@MbolotSuse I think I'm not entirely following. In this situation the resulting manifest would be an empty one. Do you mean we would like to retain the trailing `---`?

I think your question is also kind of duplicated at the end 😅

Reply:
My bad on the repeated question, sorry about that. 

In the case that I showed we would panic, since separatorIndex+apiIndex is out of bounds for the slice. In addition, I think this would lead to us chopping off more than we intended if there was a second entry after this (i.e. 51 would go into the next entry after this one).

Reply:
I tried that via a test and it didn't panic (I ended up with an empty manifest though). Did you get it to panic in any specific scenario? Could you add more details for me to investigate?

Reply:
Now that we are discussing it I believe my code does panic if the API is not the first thing in the manifest. I'll add some code for that scenario.

Reply:
I had a misunderstanding on a part - I didn't understand why we were adding the apiIndex to the separatorIndex. I got it now though - it's because we are cutting off everything before the apiIndex when we search for the next "---" occurrence. I'll resolve this for now, though I still need to take a close look at the indexes to make sure that we don't have any off-by-one errors.